aten/src/ATen/native/cuda/MultiTensorApply.cuh

```diff
@@ -38,13 +48,184 @@ __device__ __forceinline__ void load_store(
+    const std::vector<T>& vec) {
+  ptrs.push_back(vec.data());
+  // Align the offset by sizeof(T)
+  total_bytes = at::round_up(total_bytes, sizeof(T));
+  offsets.push_back(total_bytes);
+  sizes.push_back(sizeof(T) * vec.size());
+  total_bytes += sizeof(T) * vec.size();
+}
+
+// Pack multiple vectors into a DevArrayPack
+//
+// When the total size of the input vectors is large, the function packs the
+// vectors into a device buffer allocated with CUDACachingAllocator. Thanks to
+// the allocator's stream awareness, the buffer can be safely accessed by the
+// kernel as long as the tensor owning the buffer is alive at the time of the
+// kernel launch. This tensor is returned to the caller, who is responsible for
+// keeping it alive until the kernel is launched.
+template <int n, typename... Vectors>
+std::tuple<DevArrayPack, c10::optional<at::Tensor>> pack_vectors(
+    const at::Device& device,
+    std::vector<const void*> (&addresses)[n],
+    const Vectors&... vecs) {
+  std::vector<const void*> ptrs;
+  std::vector<size_t> sizes;
+  std::vector<size_t> offsets;
+  size_t total_bytes = 0;
+
+  for (auto d = 0; d < n; ++d) {
+    pack_vectors_helper(ptrs, sizes, offsets, total_bytes, addresses[d]);
+  }
+  (pack_vectors_helper(ptrs, sizes, offsets, total_bytes, vecs), ...);
+
+  TORCH_CHECK(ptrs.size() <= DevArrayPack::max_arrays);
+  DevArrayPack pack{};
+
+  // Use the small buffer in DevArrayPack to pack the vectors
+  if (total_bytes < DevArrayPack::small_buffer_size) {
+    pack.buffer_ptr = nullptr;
+    for (const auto i : c10::irange(ptrs.size())) {
+      pack.offsets[i] = offsets[i];
+      memcpy(pack.small_buffer + offsets[i], ptrs[i], sizes[i]);
+    }
+    return std::make_tuple(pack, c10::optional<at::Tensor>(c10::nullopt));
+  }
+
+  bool is_capturing = at::cuda::currentStreamCaptureStatusMayInitCtx() !=
+      at::cuda::CaptureStatus::None;
+
+  // Use dynamically allocated buffer to pack the vectors
+  // First, stage the buffer content in host memory
+  auto buf_tensor = at::empty(
+      {static_cast<int64_t>(total_bytes)},
+      // Only use pinned memory when not capturing
+      at::TensorOptions().dtype(at::kByte).pinned_memory(!is_capturing));
+  for (const auto i : c10::irange(ptrs.size())) {
+    pack.offsets[i] = offsets[i];
+    memcpy(buf_tensor.data_ptr<uint8_t>() + offsets[i], ptrs[i], sizes[i]);
+  }
+
+  if (is_capturing) {
+    // The dynamically allocated buffer can be considered an extension of the
+    // kernel argument. Therefore, when used with CUDA graph, the buffer should
+    // remain constant across replays and have the same lifetime as the
+    // captured graph. By allocating the buffer via
+    // CUDACachingAllocator::raw_alloc and allowing it to "leak", a region in
+    // the CUDA graph's pool is reserved for the duration of the graph's
+    // lifetime.
+    pack.buffer_ptr = static_cast<char*>(
+        c10::cuda::CUDACachingAllocator::raw_alloc(total_bytes));
```

Comment:
If you do raw_alloc you're responsible for calling `raw_delete` after. Could we just allocate a tensor or storage here ? If we're cudagraphing, there is no C++ overhead so no need to optimize for that.

Reply:
The intention is not to optimize for CPU overhead, but to mimic the kernel argument semantics of CUDA graph.

If I understand correctly, CUDA graph allocates device memory for kernel arguments and copies them from host to device at capture time. The kernel arguments on device have the same lifetime as the CUDA graph.

Since the high level idea of this PR is to improve `multi_tensor_apply`'s efficiency by not spliting workloads due to the 4kb limitation of the static kernel argument space, the buffer in this context effectively serves as an extension of the kernel argument. Therefore, I thought it might be acceptable to not invoke `raw_delete` here, similar to how regular kernel arguments are never deleted. What do you think?

Reply:
Why would a regular ref-counted style allocation not work here? If it's during a graph capture, it should happen on the private memory pool, and while it doesn't look like a huge amount of memory the caching allocator would be free to reclaim it for other purposes once this operation is finished. Note that the requirement is _not_ that the buffer has to be alive for the lifetime of the graph, but rather that each replay can reuse the same address, and this is exactly the use-case that the private memory pool functionality of the caching allocator is supposed to support.

A side question, if this `raw_alloc` happens during a capture on a private pool, would we leak the memory indefinitely, or only until the graph is destroyed?

Reply:
> Note that the requirement is not that the buffer has to be alive for the lifetime of the graph, but rather that each replay can reuse the same address, and this is exactly the use-case that the private memory pool functionality of the caching allocator is supposed to support.

Thanks for your input, @eqy. In this context, the buffer holds metadata such as pointers and sizes, as opposed to the type of data typically held by tensors. Therefore, I believe it makes more sense to bake the content of the buffer into the graph. The `raw_alloc` combined with capture-time only `cudaMemcpy` aims to achieve this. Consider the following example:

```cpp
constexpr size_t N = ...;
struct Foo {
  int64_t* pointers[N];
}
__global__ my_kernel(Foo foo) {...}
```

During capture, CUDA graph copies the content of `Foo` onto the device, where it remains for the lifetime of the graph. Now, suppose we want to lift the 4KB size limit of `Foo` by replacing pointers with a dynamically allocated buffer (which is effectively the approach used by this PR). I believe it would be reasonable for the buffer to receive the same treatment as `Foo` - reserved for the duration of the graph's lifetime and populated once during capture.

> Why would a regular ref-counted style allocation not work here?

A ref-counted allocation implies that even though we get a baked-in address, the region pointed to by the address could be reassigned for other purposes, leading to the buffer's content being overwritten.

In theory, we could use ref-counted style allocation here and populate the baked-in device location with a HtoD copy on every replay. However, this is unnecessary and inefficient in this context, and I'm uncertain whether HtoD copies can be recorded without direct graph manipulation.

> A side question, if this raw_alloc happens during a capture on a private pool, would we leak the memory indefinitely, or only until the graph is destroyed?

IIUC only until the graph is destroyed.

--------

To summarize, there are two options for this buffer:
- reserve it for the duration of the graph's lifetime and populate it once during capture
- populate its content with a HtoD copy on every replay

I believe the first option makes more sense. Let me know what you think.

Reply:
I dont think we are actually skipping the cudaMemcpy currently despite the `raw_alloc`. You can verify with `debug_dump()` in Graph.cpp. 

In either case - we should could a benchmarking demonstrating the win here.

Reply:
> I dont think we are actually skipping the cudaMemcpy currently despite the raw_alloc. You can verify with debug_dump() in Graph.cpp.

The cudaMemcpy is skipped according to the dump. 

```
> digraph dot {
subgraph cluster_1 {
label="graph_1" graph[style="dashed"];
"graph_1_node_0"[style="bold" shape="octagon" label="0
_ZN2at6native55_GLOBAL__N__791b5f0c_22_ForeachBinaryOpList_cu_a911ec4325multi_tensor_apply_kernelINS1_18TensorListMetadataILi2EEENS1_14UnaryOpFunctorIfLi2ELi1ELi1EEEJNS0_8IdentityIfEEEEEvNS1_12DevArrayPackET0_DpT1_
node handle: 0x00000000086ED680
func handle: 0x00000000085718D0
"];

}
}
```

Also, according to https://www.nvidia.com/en-us/on-demand/session/gtcspring23-s51211/, the non-async version of cudaMemcpy is not capturable.

<img width="764" alt="image" src="https://github.com/pytorch/pytorch/assets/4156752/ab18df52-8769-4f7e-bb0c-529be2c0c19f">


> In either case - we should could a benchmarking demonstrating the win here.

I'm not sure whether the alternative, i.e. recording a HtoD copy can be achievied via stream capture (no direct graph manipulation). Do you know?

Reply:
In that case, what prevents us from giving the optimizer a (non-parameter) tensor member (of e.g., dtype `bytes`), that is allocated on the default memory pool and only created when e.g., a `use_graphs` argument is set to `True` similar to `capturable` here: https://github.com/pytorch/pytorch/blob/c802c5019602d5ae6e43f8079d6bbad48a53ef60/torch/optim/adam.py#L25 ?

Reply:
@eqy thanks for offering the idea. It is indeed a good pattern for this type of problems in general.

However, the main challenge of applying it here is that not all upstream users of `multi_tensor_apply` are stateful like optimizers. `foreach` kernels are also used without optimizers in many cases. Requring users to maintain and pass in a "workspace" buffer to make any `foreach` op capturable seems like a degradation in UX.

We can give this approach more thought. However, do you see any substantial downsides to the proposed approach, aside from it being a bit unconventional? It only appears that we are leaking memory, but this is not actually the case since the lifetime of the allocation is tied to the graph.

Reply:
I think the reasons I am giving are purely aesthetic, mainly that this extra buffer seems to be a special case of multi-tensor apply, and intuitively the "special-case" handling might make more sense to have in the thing that actually is the special case rather than the generic CUDA graphs code.

I agree this is justifiable to enable e.g., the non-optimizer use cases of `multi_tensor_apply` as you mentioned.

Reply:
I echo your point on aesthetics. I updated the PR to use CUDA User Object to manage the lifetime of the extended launch arguments. Hopefully it addresses the concern.